### DISCUSSION TITLE
Parse errors when conditional directives are combined with raw string literals

### DISCUSSION THREAD

**Discussion by Youssef1313** [dotnet/csharplang#7520] [General]
```csharp
#if false

var x = """
    # Hello raw string literals
    """;
    
#endif
```

I can see why that fails, but I'm wondering if this can be improved. I think that when the compiler encounters a conditional directive that evaluates to false, it should try to keep going until it finds a valid directive trivia.

So instead of parsing "# Hello raw string literals" with as "BadDirectiveTrivia", it should be parsed as DisabledTextTrivia and continue searching for a valid directive trivia, which is going to be #endif

cc @CyrusNajmabadi

**Comment by Youssef1313** [comment]
This also happens with the verbatim strings.

The ambiguity will still be there anyway if the verbatim/raw string contains syntactically valid directives though :/

**Comment by CyrusNajmabadi** [comment]
This is a request for a language change.  The compiler is working as per the language rules here.  I'm wary about any changes here because of hte potential for regressions.

**Comment by Youssef1313** [comment]
Is the current behavior well-spec'ed? If so, can you move this as a discussion to csharplang?

**Comment by CyrusNajmabadi** [comment]
> Is the current behavior well-spec'ed? If so, can you move this as a discussion to csharplang?

Yes, it's spec'ed here: https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/lexical-structure.md#65-pre-processing-directives

**Comment by TahirAhmadov** [comment]
I'm confused now. From https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/lexical-structure.md#65-pre-processing-directives

![image](https://github.com/dotnet/csharplang/assets/16783987/ecfe592b-4647-42ea-ad44-9360fa148b8d)

It sounds like "pre-processing directives" are not processed, but are still parsed. Basically, you cannot have a line inside any multi-line string which starts with a "#", except if it happens to be valid directive.

If this is correct, then:
- As implemented today, this is a significant design flaw - a needless limitation - frankly, I'm surprised many more haven't noticed this until now;
- The spec is not clear; it says they are not processed, but does not specifically address whether they are parsed (as implemented today, they are);
- Changing the behavior here cannot introduce regressions, because as specced, they are not processed already, and if we make the change (in spec and roslyn) to not parse them anymore, it makes a previously illegal scenario legal, which is a viable change.

Or am I missing something?

**Reply by TahirAhmadov** [reply]
OK never mind, I see what is happening now. It's only inside inactive `#if` blocks.
To enable this code to be legal, the compiler would need to somehow partially compile what's inside an inactive `#if` block, which defeats the purpose of it being inactive.

**Comment by Unknown6656** [comment]
[Same happens with block comments:](https://sharplab.io/#v2:CYLg1APgsAUAxASwGYAICEAXATgVwKawpEqwBuAhligB4oC8KARM4cYgHYUA2CwNszRgG5WRMpRQBPeigACsxqJQduvKUuJMRMTZtgB6AFRKV5HnwBesQ/thw87YMiA=)
![image](https://github.com/dotnet/csharplang/assets/8807985/7ac764eb-04fc-46cd-9982-f66e1dfe9db1)

**Reply by HaloFour** [reply]
That's what I'd expect, the compiler isn't attempting to parse any code within the block.  It is still looking for preprocessor directives, though.

**Reply by CyrusNajmabadi** [reply]
Indeed. This is all by design, and how things have worked since c#1.0.

**Reply by TahirAhmadov** [reply]
Is it possible to change the spec so that a line inside in an inactive `#if` block which starts with `#` but is not a valid pre-processor directive, is summarily ignored?

**Reply by 333fred** [reply]
That would then be a breaking change if we ever want to add a directive with that text.
